### PR TITLE
feat: paymaster feature

### LINKED ISSUE #826: Prepare `freeregistration.tsx` for Ethcc
## Description 

There are several practical features we want to add to the `freeregistration.tsx` for ethcc

- [x] Make this link https://app.starknet.id/gift open the Argent/Braavos starknet id page (you are in the telegram group with both wallets and we asked them how to do that). 
- [x] https://github.com/starknet-id/app.starknet.id/issues/827

- [ ] Add paymaster feature

Paymaster is a feature where the Starknet Foundation can subsidize user TX fees when they use a dApp. To be able to make the tx free (if the user is eligible) we need to add the implementation on our front end (in our case on `freeregistration.tsx`). We'll not manage the process of making a user eligible for the moment, we just want to check: 
1. If the user is eligible for free tx
2. If yes we then ask to check if the user has a deployed wallet
3. And then we'll build and send the tx based on the previous response (with or without paymaster, with or without wallet deployment in the paymaster tx).

[Here is the guide on how to do it using the AVNU paymaster api.](https://starknet.notion.site/GUIDE-Paymaster-Protocol-Integration-d1fd1d963c6e448997542c4d1c1c03fd)

It can be a tricky PR so feel free to ask any question on the Core tribe group, the Avnu group or the wallets group.

### PR DESCRIPTION
**Pull request by Marchand-Nicolas**
close: #826 
To test, go to /gift 

## Summary by CodeRabbit

- **New Features**
  - Introduced `usePaymaster` custom React hook to manage gasless transactions, including account checks, gas token pricing, rewards, and transaction execution.
  - Added `tokenNames` function for mapping token addresses to their names.

- **Updates**
  - Updated multiple dependencies in `package.json`:
    - Added `@avnu/gasless-sdk` and `qs`.
    - Updated `@starknet-react/chains`, `@starknet-react/core`, and `get-starknet-core`.

- **Enhancements**
  - Revised `FreeRegisterCheckout` component to incorporate `usePaymaster`, enhancing the handling of paymaster-related functionalities and gas token management.

### CODE REVIEW

File: components/UI/navbar.tsx [hunk 1]
```diff
@@ -9,14 +9,15 @@ import { AiOutlineMenu } from "react-icons/ai";
 import { FaDiscord, FaGithub, FaTwitter } from "react-icons/fa";
 import styles from "../../styles/components/navbar.module.css";
 import connectStyles from "../../styles/components/walletConnect.module.css";
+
```
Comment (fricoben) [inline components/UI/navbar.tsx]: delete

File: components/UI/navbar.tsx [hunk 2]
```diff
@@ -53,7 +55,6 @@ const Navbar: FunctionComponent = () => {
   const [showWalletConnectModal, setShowWalletConnectModal] =
     useState<boolean>(false);
   const [lastConnector, setLastConnector] = useState<Connector | null>(null);
-
```
Comment (fricoben) [inline components/UI/navbar.tsx:56]: add back the jump of line

File: components/UI/navbar.tsx [hunk 3]
```diff
@@ -87,15 +88,12 @@ const Navbar: FunctionComponent = () => {
   useEffect(() => {
     address ? setIsConnected(true) : setIsConnected(false);
   }, [address]);
-
   useEffect(() => {
```
Comment (fricoben) [inline components/UI/navbar.tsx]: add back the jump of line

File: utils/callData/registrationCalls.ts [hunk 4]
```diff
@@ -84,7 +84,7 @@ function mint(tokenId: number): Call {
   return {
     contractAddress: process.env.NEXT_PUBLIC_IDENTITY_CONTRACT as string,
     entrypoint: "mint",
-    calldata: [numberToString(tokenId)],
+    calldata: [numberToStringHex(tokenId)],
```
Comment (fricoben) [inline utils/callData/registrationCalls.ts:87]: Why are you changing here ?

File: utils/callData/registrationCalls.ts [hunk 5]
```diff
@@ -221,7 +221,12 @@ export const getFreeRegistrationCalls = (
     {
       contractAddress: process.env.NEXT_PUBLIC_DOMAIN_GIFT_CONTRACT as string,
       entrypoint: "get_free_domain",
-      calldata: [newTokenId, encodedDomain, signature, txMetadataHash].flat(),
+      calldata: [
+        numberToStringHex(newTokenId),
+        numberToStringHex(encodedDomain),
+        signature.map((s) => numberToStringHex(s)),
+        txMetadataHash,
+      ].flat(),
```
Comment (fricoben) [inline utils/callData/registrationCalls.ts:229]: You you change here ?

File: components/discount/freeRegisterCheckout.tsx [hunk 6]
```diff
@@ -59,7 +95,110 @@ const FreeRegisterCheckout: FunctionComponent<FreeRegisterCheckoutProps> = ({
...
+    });
+  }, [
+    gasMethod,
+    callData,
+    account,
+    gasTokenPrice,
+    gaslessCompatibility,
+    estimateCalls,
+    gaslessAPIAvailable,
+  ]);
```
Comment (fricoben) [inline components/discount/freeRegisterCheckout.tsx:88]: Seems big chunk of code. 

Could you make it a separate hook especially if we can reuse this elsewhere in another registration flow ...

**Review by fricoben — CHANGES_REQUESTED** [review]
## Code review

Some code comments, I just checked the code caused I don't know how to test E2E for the moment

File: utils/callData/registrationCalls.ts (hunk 4 above)
Comment (Marchand-Nicolas) [reply utils/callData/registrationCalls.ts:87]: Paymaster returns a 500 error when we use decimal numbers, so I chose to use hexadecimals as they are supported by both wallet tx and paymaster

File: utils/callData/registrationCalls.ts (hunk 5 above)
Comment (Marchand-Nicolas) [reply utils/callData/registrationCalls.ts:229]: Same as here https://github.com/starknet-id/app.starknet.id/pull/841#discussion_r1655224540

**Review by fricoben — APPROVED** [review]
lgtm

**Pull request merged**